### PR TITLE
Add an abstract interpreter 'history' that gives (local) per-path analysis and / or loop unwinding

### PR DESCRIPTION
**Pull request by martin-cs**
In the past users have had to modify the program before analysis to get this kind of precision.  Doing so is expensive in terms of memory for both the program and the analysis.  This gives a more efficient route.  The is an equivalent of an unwinding limit that controls the number of histories that can be generated for each location.  This is a better measure as it should scale memory linearly and time linearly(ish) while loop unwinding can have seriously non-linear effects when there are nested loops.

This is the enabling technology for a "loop unwinding" analysis that can automatically over-approximate loop bounds.  It is also probably the last "core" change to the AI infrastructure before VSD can be merged in full.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by martin-cs** [comment]
Please can someone with access to the AWS CodeBuild outputs let me know what the problem is?  Last time it was @chrisr-diffblue or @thk123 who could do this...

**Comment by chrisr-diffblue** [comment]
Looks like all of the failed AWS codebuild jobs failed really early due to infrastructure problems (failed to apt-get install dependencies). I've restarted all those failed jobs.

**Comment by martin-cs** [comment]
Thanks @chrisr-diffblue  they are listed as "failed" rather than "pending" for me at the moment.  I'll give them a minute...

**Comment by martin-cs** [comment]
These still look failed to me, plus the code coverage numbers are kinda wacky.

**Comment by chrisr-diffblue** [comment]
Unfortunately, it looks like although the restarted jobs seem to have passed, they haven't posted status back to github, sigh. You might have to go for the rebase/dummy commit + force push to make github start new jobs.

**Comment by smowton** [comment]
@chrisr-diffblue IIRC updating Github status after a rerun is a Codebuild option, @rjmunro knows more

**Comment by rjmunro** [comment]
@smowton @chrisr-diffblue Yes, there is an option "Report build statuses to source provider when your builds start and finish" in the "Edit" -> "Source" menu, hidden in "Additional configuration (Git clone depth, Git submodules)" (even though it is neither of those things). Without this, codebuild will report status to github only where the build was triggered by a github webhook. I can see it's not turned on, but I'd rather not change the setting myself as I'm not involved in this project. I'm happy to do it if someone officially asks me to.

Finding that setting was a huge win for us because it let us trigger codebuild jobs from other codebuild jobs, rather than from webhooks. This meant we could build a CI pipeline.

**Comment by thk123** [comment]
@rjmunro I'm happy to do it - any  reason to _not_ have this on? Seems like a no-brainer...

**Comment by rjmunro** [comment]
@thk123 The guess the main reason not to have it on would be if you were testing your codebuild config and didn't want failures to block PRs or something, but in general, I would turn it on.

**Comment by martin-cs** [comment]
Thanks for the help all but ... I still can't see what is actually causing the failures.  I'd really like to get this PR moving and I'm keen to fix the issues but without the info...

**Comment by thk123** [comment]
@martin-cs as Chris said - the failures seem spurious - you should just be able to rebase this and push to retrigger the builds. I'll try and get the updated status thing working today.

**Comment by martin-cs** [comment]
Thank you @thk123 that seems to have done the trick.

Do any of the maintainers have time for a review of this PR?

File: src/analyses/ai.cpp
Comment (smowton) [inline src/analyses/ai.cpp]: This and other `nullptr` args are pretty opaque -- how about `#define CALLER_UNSPECIFIED nullptr`?

File: src/analyses/ai.cpp
Comment (smowton) [inline src/analyses/ai.cpp]: Why is this one not given?

File: src/analyses/ai.h
Comment (smowton) [inline src/analyses/ai.h]: Does the nullptr here imply we're conflating some states that we could potentially distinguish?

File: src/analyses/local_control_flow_history.h
Comment (smowton) [inline src/analyses/local_control_flow_history.h:65]: POSTCONDITION?

File: src/analyses/local_control_flow_history.h
Comment (smowton) [inline src/analyses/local_control_flow_history.h]: rename `to_local_control_flow_history`?

File: src/analyses/local_control_flow_history.cpp
Comment (smowton) [inline src/analyses/local_control_flow_history.cpp:272]: Looks like you intentionally want more informative orderings to be lower in the sort order? If so say so up top

File: src/analyses/local_control_flow_history.cpp
Comment (smowton) [inline src/analyses/local_control_flow_history.cpp:359]: Being all possible versions :)

File: src/goto-analyzer/goto_analyzer_parse_options.cpp
Comment (smowton) [inline src/goto-analyzer/goto_analyzer_parse_options.cpp:301]: Considering there's some trickiness here, perhaps print a line describing the options you've inferred from the higher-level options provided

File: src/goto-analyzer/goto_analyzer_parse_options.cpp
Comment (smowton) [inline src/goto-analyzer/goto_analyzer_parse_options.cpp:308]: The else-ifs here imply `--loop-unwind --branching` won't work as expected, right? If so check for this case.

**Review by smowton — APPROVED** [review]
Mostly lgtm, some requests for clarity, particularly clarifying either by naming a pass-through function or using a named `nullptr` when histories are being omitted because they're irrelevant, and when they're being omitted due to a strategic choice to ignore history and merge all states.

File: src/analyses/ai.cpp
Comment (hannes-steffenhagen-diffblue) [reply src/analyses/ai.cpp]: I don't really agree with using a define for this. How about `constexpr whatever* UnspecifiedCaller = nullptr;`?

File: src/analyses/ai.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/ai.cpp:325]: I do have a pretty strong preference for not using `nullptr` to indicate optional arguments – this isn't obvious from the signature, and frequently makes the logic inside more complicated if you need to disambiguate between pointers you're allowed to dereference and ones you aren't...

Also, are these `trace_ptrts` all `shared_ptr`? In that case it'd seem a bit wasteful to me,  unless these are stored somewhere (I can't quite see it, maybe somewhere further down in the call stack)?

File: src/analyses/ai.cpp
Comment (martin-cs) [reply src/analyses/ai.cpp]: Added to ai_historyt as a member variable as @hannes-steffenhagen-diffblue suggests.

File: src/analyses/ai.cpp
Comment (martin-cs) [reply src/analyses/ai.cpp]: Because it is only used on function call return edges, this is a call, so `p_call` already has the caller info.

File: src/analyses/ai.h
Comment (martin-cs) [reply src/analyses/ai.h]: Good question but in this case no.  The step is local so the history is not needed.  I have added a comment to this effect.

( There is a separate conversation about whether such an interface is even meaningful w.r.t. function calls and goto instructions. )

File: src/analyses/local_control_flow_history.h
Comment (martin-cs) [reply src/analyses/local_control_flow_history.h:65]: I think it is a precondition of `down_cast` that this conversion works.  To quote the comment above `caller's responsibility to ensure correct types`.

File: src/analyses/local_control_flow_history.h
Comment (martin-cs) [reply src/analyses/local_control_flow_history.h]: Done.

File: src/analyses/local_control_flow_history.cpp
Comment (martin-cs) [reply src/analyses/local_control_flow_history.cpp:272]: Ummm... I am not entirely sure I have an idea of what constitutes informative in this context.  Where the choice of ordering matters (both in this and in `local_control_flow_historyt<track_forward_jumps, track_backward_jumps>::
operator<`) it is commented and in one case checked by assertion (if there is a merge history it should sort last).  Specifically this includes:
```
// Lower locations first, generally this means depth first
// Branch not taken takes priority
// This prioritises complete histories over merged ones
```

File: src/analyses/local_control_flow_history.cpp
Comment (martin-cs) [reply src/analyses/local_control_flow_history.cpp:359]: That too!  (although now I think about it maybe the version that doesn't track either isn't actually used... )

File: src/goto-analyzer/goto_analyzer_parse_options.cpp
Comment (martin-cs) [reply src/goto-analyzer/goto_analyzer_parse_options.cpp:301]: The options that are set are internal only and are intended to make the handling of these options simpler.  I am reluctant to print anything here because the user can't set any of these options directly.

File: src/goto-analyzer/goto_analyzer_parse_options.cpp
Comment (martin-cs) [reply src/goto-analyzer/goto_analyzer_parse_options.cpp:308]: That is correct and in part why the combined option is known as `--loop-unwind-and-branching`.  I am reluctant to address this issue in this PR because it affects almost all options to goto-analyze`, all of the major options have an implicit ordering and check for absence but not duplication.

File: src/analyses/ai.cpp
Comment (martin-cs) [reply src/analyses/ai.cpp:325]: I have renamed as you suggested earlier.  I hope the invariants in `ai_history.h` show when the optional argument is present.  `step` kind of has to be a case switch because of the different types of instruction.

Yes, they are all `shared_ptr`s.  They are stored in something that implements the `ai_storaget` interface, allowing you to store states and histories with varying degrees of sensitivity (per history, per location, soon per function and just once).  It's not great but I had a solution with `unique_ptr` but it involved way too many copies.

File: src/analyses/ai.cpp
Comment (martin-cs) [reply src/analyses/ai.cpp]: ( Unfortunately `constexpr` doesn't actually work because `ai_history_baset` has a virtual destructor but a `static const` variable does. )

**Comment by martin-cs** [comment]
@smowton and @hannes-steffenhagen-diffblue I have responded to and (mostly) implemented your review comments.  They are two separate commits at the moment for simplicity.  If you are happy I will squash them into the relevant earlier commits.

**Comment by martin-cs** [comment]
@smowton and @hannes-steffenhagen-diffblue sorry to hassle but please can you have a look at the changes I have made and see if they address the issues you raised?

**Comment by smowton** [comment]
I stand by my green tick :) (meaning I'm happy with this as-is but have suggestions)

**Comment by smowton** [comment]
Apologies for slowness btw, was on holiday the past week

**Comment by martin-cs** [comment]
@smowton Thanks!  Hope you had a nice holiday and sorry for hassling.

@hannes-steffenhagen-diffblue @danpoe @peterschrammel @chrisr-diffblue Any chance you could help with a second green tick?  This is kind of blocking for me and the next PR is something that might make your lives easier...

File: src/analyses/call_stack_history.cpp
Comment (chrisr-diffblue) [inline src/analyses/call_stack_history.cpp]: Can you use `ai_history_baset::no_caller_history` rather in this INVARIANT, rather than `nullprt`?

File: src/analyses/local_control_flow_history.cpp
Comment (chrisr-diffblue) [inline src/analyses/local_control_flow_history.cpp]: Ditto for the `nullptr` here too

**Review by chrisr-diffblue — APPROVED** [review]
A couple of extremely minor nitpicks, but nothing blocking - apologies for the massive delay in me getting around to reviewing....

File: src/analyses/call_stack_history.cpp
Comment (martin-cs) [reply src/analyses/call_stack_history.cpp]: Done.

File: src/analyses/local_control_flow_history.cpp
Comment (martin-cs) [reply src/analyses/local_control_flow_history.cpp]: Done.

**Comment by martin-cs** [comment]
I have addressed @chrisr-diffblue comments ( thanks for the review and no worries about the lag ) and squashed the changes in.  I will merge once CI has passed, @hannes-steffenhagen-diffblue shout now if you aren't happy with how your comments have been resolved.

**Pull request merged**
